// storybook@10.2.19/dist/telemetry/index.js
import CJS_COMPAT_NODE_URL_q3hucshmk7 from 'node:url';
import CJS_COMPAT_NODE_PATH_q3hucshmk7 from 'node:path';
import CJS_COMPAT_NODE_MODULE_q3hucshmk7 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_q3hucshmk7.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_q3hucshmk7.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_q3hucshmk7.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  ErrorCollector,
  addToGlobalContext,
  cleanPaths,
  computeStorybookMetadata,
  getLastEvents,
  getPrecedingUpgrade,
  getSessionId,
  getStorybookMetadata,
  isExampleStoryId,
  metaFrameworks,
  oneWayHash,
  removeAnsiEscapeCodes,
  sanitizeAddonName,
  sanitizeError,
  telemetry
} from "../_node-chunks/chunk-XRGW44MZ.js";
import "../_node-chunks/chunk-RFX2XUEG.js";
import "../_node-chunks/chunk-2RKLK4B2.js";
import "../_node-chunks/chunk-UCFQGJDO.js";
import "../_node-chunks/chunk-7L22HHNN.js";
import "../_node-chunks/chunk-V5TALVQU.js";
import "../_node-chunks/chunk-SBCV2VV5.js";
import "../_node-chunks/chunk-HBF5DW76.js";
import "../_node-chunks/chunk-DAUUD63K.js";
import "../_node-chunks/chunk-CGEEACTP.js";
import "../_node-chunks/chunk-JGHDDUW7.js";
import "../_node-chunks/chunk-HIZAGWGK.js";
import "../_node-chunks/chunk-UMKJ2KQ3.js";
import "../_node-chunks/chunk-57ELZD3S.js";
import "../_node-chunks/chunk-FJYNYOAM.js";
import "../_node-chunks/chunk-ADTXI3R5.js";
import "../_node-chunks/chunk-GD5QW2LP.js";
export {
  ErrorCollector,
  addToGlobalContext,
  cleanPaths,
  computeStorybookMetadata,
  getLastEvents,
  getPrecedingUpgrade,
  getSessionId,
  getStorybookMetadata,
  isExampleStoryId,
  metaFrameworks,
  oneWayHash,
  removeAnsiEscapeCodes,
  sanitizeAddonName,
  sanitizeError,
  telemetry
};
