// storybook@10.2.19/dist/cli/index.js
import CJS_COMPAT_NODE_URL_q3hucshmk7 from 'node:url';
import CJS_COMPAT_NODE_PATH_q3hucshmk7 from 'node:path';
import CJS_COMPAT_NODE_MODULE_q3hucshmk7 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_q3hucshmk7.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_q3hucshmk7.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_q3hucshmk7.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  ANGULAR_JSON_PATH,
  AddonVitestService,
  AngularJSON,
  ProjectType,
  SUPPORTED_ESLINT_EXTENSIONS,
  addToDevDependenciesIfNotPresent,
  adjustTemplate,
  cliStoriesTargetPath,
  coerceSemver,
  configureEslintPlugin,
  configureFlatConfig,
  copyTemplate,
  copyTemplateFiles,
  extractEslintInfo,
  findEslintFile,
  getBabelDependencies,
  getRendererDir,
  getVersionSafe,
  hasStorybookDependencies,
  normalizeExtends,
  readFileAsJson,
  suggestESLintPlugin,
  writeFileAsJson
} from "../_node-chunks/chunk-M3DDSQUK.js";
import {
  detectPnp
} from "../_node-chunks/chunk-XKYZMTQG.js";
import {
  Settings,
  _clearGlobalSettings,
  globalSettings
} from "../_node-chunks/chunk-HBF5DW76.js";
import "../_node-chunks/chunk-DAUUD63K.js";
import "../_node-chunks/chunk-CGEEACTP.js";
import "../_node-chunks/chunk-JGHDDUW7.js";
import "../_node-chunks/chunk-HIZAGWGK.js";
import "../_node-chunks/chunk-FJYNYOAM.js";
import "../_node-chunks/chunk-OOCT2CCT.js";
import "../_node-chunks/chunk-R37DSTA2.js";
import "../_node-chunks/chunk-ADTXI3R5.js";
import "../_node-chunks/chunk-GD5QW2LP.js";
export {
  ANGULAR_JSON_PATH,
  AddonVitestService,
  AngularJSON,
  ProjectType,
  SUPPORTED_ESLINT_EXTENSIONS,
  Settings,
  _clearGlobalSettings,
  addToDevDependenciesIfNotPresent,
  adjustTemplate,
  cliStoriesTargetPath,
  coerceSemver,
  configureEslintPlugin,
  configureFlatConfig,
  copyTemplate,
  copyTemplateFiles,
  detectPnp,
  extractEslintInfo,
  findEslintFile,
  getBabelDependencies,
  getRendererDir,
  getVersionSafe,
  globalSettings,
  hasStorybookDependencies,
  normalizeExtends,
  readFileAsJson,
  suggestESLintPlugin,
  writeFileAsJson
};
