// storybook@10.2.19/dist/node-logger/index.js
import CJS_COMPAT_NODE_URL_q3hucshmk7 from 'node:url';
import CJS_COMPAT_NODE_PATH_q3hucshmk7 from 'node:path';
import CJS_COMPAT_NODE_MODULE_q3hucshmk7 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_q3hucshmk7.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_q3hucshmk7.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_q3hucshmk7.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  CLI_COLORS,
  ConsoleLogger,
  StyledConsoleLogger,
  colors,
  createHyperlink,
  deprecate,
  import_npmlog,
  logTracker,
  logger,
  once,
  prompt,
  protectUrls
} from "../_node-chunks/chunk-V5TALVQU.js";
import "../_node-chunks/chunk-ADTXI3R5.js";
import "../_node-chunks/chunk-GD5QW2LP.js";
var export_instance = import_npmlog.default;
export {
  CLI_COLORS,
  ConsoleLogger,
  StyledConsoleLogger,
  colors,
  createHyperlink,
  deprecate,
  export_instance as instance,
  logTracker,
  logger,
  once,
  prompt,
  protectUrls
};
