// vite-plus@0.1.20/dist/migration/bin.js
import { r as __toESM } from "../chunk-q7NCDQ7-.js";
import { l as isForceOverrideMode } from "../main-A6UrSTYb.js";
import { a as printHeader, i as muted, r as log, t as accent } from "../terminal-CxTMfsxZ.js";
import { $ as outro, A as warnLegacyEslintConfig, B as runViteInstall, C as preflightGitHooksSetup, E as rewriteMonorepo, F as cancelAndExit, H as upgradeYarn, I as defaultInteractive, K as PackageManager, L as downloadPackageManager, M as warnPackageLevelPrettier, O as rewriteStandaloneProject, R as promptGitHooks, S as migratePrettierToOxfmt, T as promptPrettierMigration, U as displayRelative, V as selectPackageManager, Y as confirm, Z as log$1, a as writeAgentInstructions, b as migrateEslintToOxlint, c as checkVitestVersion, d as detectEslintProject, et as select, f as detectFramework, g as hasFrameworkShim, it as q, j as warnPackageLevelEslint, l as confirmEslintMigration, m as detectPrettierProject, n as detectExistingAgentTargetPaths, o as addFrameworkShim, p as detectNodeVersionManagerFile, q as require_semver, r as selectAgentTargetPaths, s as checkViteVersion, t as detectAgentConflicts, tt as spinner, u as confirmPrettierMigration, v as installGitHooks, w as promptEslintMigration, x as migrateNodeVersionManagerFile, y as mergeViteConfigFiles } from "../agent-D_WSpD0r.js";
import { t as lib_default } from "../lib-BamM40b7.js";
import { a as readNearestPackageJson, i as hasVitePlusDependency } from "../package-D_LD1iiI.js";
import { r as createMigrationReport } from "../report-DbrfjWiP.js";
import { c as writeEditorConfigs, i as detectEditorConflicts, o as selectEditor, t as detectWorkspace } from "../workspace-DVOyShUK.js";
import { t as renderCliDoc } from "../help-BtkjXtRM.js";
import path from "node:path";
import { styleText } from "node:util";
//#region src/migration/bin.ts
var import_semver = /* @__PURE__ */ __toESM(require_semver(), 1);
async function confirmNodeVersionFileMigration(interactive, detection) {
	const message = {
		"package.json": "Migrate Volta node version (package.json) to .node-version?",
		".nvmrc": "Migrate .nvmrc to .node-version?"
	}[detection.file];
	if (interactive) {
		const confirmed = await confirm({
			message,
			initialValue: true
		});
		if (q(confirmed)) cancelAndExit();
		return confirmed;
	}
	return true;
}
async function confirmFrameworkShim(framework, interactive) {
	const name = {
		vue: "Vue",
		astro: "Astro"
	}[framework];
	if (interactive) {
		const confirmed = await confirm({
			message: `Add TypeScript shim for ${name} component files (*.${framework})?\n  ` + styleText("gray", `Lets TypeScript recognize .${framework} files until vp check fully supports them.`),
			initialValue: true
		});
		if (q(confirmed)) cancelAndExit();
		return confirmed;
	}
	return true;
}
const helpMessage = renderCliDoc({
	usage: "vp migrate [PATH] [OPTIONS]",
	summary: "Migrate standalone Vite, Vitest, Oxlint, Oxfmt, and Prettier projects to unified Vite+.",
	documentationUrl: "https://viteplus.dev/guide/migrate",
	sections: [
		{
			title: "Arguments",
			rows: [{
				label: "PATH",
				description: "Target directory to migrate (default: current directory)"
			}]
		},
		{
			title: "Options",
			rows: [
				{
					label: "--agent NAME",
					description: "Write coding agent instructions to AGENTS.md, CLAUDE.md, etc."
				},
				{
					label: "--no-agent",
					description: "Skip writing coding agent instructions"
				},
				{
					label: "--editor NAME",
					description: "Write editor config files into the project."
				},
				{
					label: "--no-editor",
					description: "Skip writing editor config files"
				},
				{
					label: "--hooks",
					description: "Set up pre-commit hooks (default in non-interactive mode)"
				},
				{
					label: "--no-hooks",
					description: "Skip pre-commit hooks setup"
				},
				{
					label: "--no-interactive",
					description: "Run in non-interactive mode (skip prompts and use defaults)"
				},
				{
					label: "-h, --help",
					description: "Show this help message"
				}
			]
		},
		{
			title: "Examples",
			lines: [
				`  ${muted("# Migrate current package")}`,
				`  ${accent("vp migrate")}`,
				"",
				`  ${muted("# Migrate specific directory")}`,
				`  ${accent("vp migrate my-app")}`,
				"",
				`  ${muted("# Non-interactive mode")}`,
				`  ${accent("vp migrate --no-interactive")}`
			]
		},
		{
			title: "Migration Prompt",
			lines: [
				"  Give this to a coding agent when you want it to drive the migration:",
				"",
				"  Migrate this project to Vite+.",
				"  Vite+ replaces the current split tooling around runtime management, package",
				"  management, dev/build/test commands, linting, formatting, and packaging.",
				"  Run `vp help` and `vp help migrate` before making changes.",
				`  Use ${accent("vp migrate --no-interactive")} in the workspace root.`,
				"  Make sure the project is using Vite 8+ and Vitest 4.1+ before migrating.",
				"",
				"  After the migration:",
				"  - Confirm `vite` imports were rewritten to `vite-plus` where needed",
				"  - Confirm `vitest` imports were rewritten to `vite-plus/test` where needed",
				"  - Remove old `vite` and `vitest` dependencies only after those rewrites",
				"    are confirmed",
				"  - Move remaining tool-specific config into the appropriate blocks in",
				"    `vite.config.ts`",
				"",
				"  Command mapping:",
				"  - `vp run <script>` is the equivalent of `pnpm run <script>`",
				"  - `vp test` runs the built-in test command, while `vp run test` runs the",
				"    `test` script from `package.json`",
				"  - `vp install`, `vp add`, and `vp remove` delegate through the package",
				"    manager declared by `packageManager`",
				"  - `vp dev`, `vp build`, `vp preview`, `vp lint`, `vp fmt`, `vp check`,",
				"    and `vp pack` replace the corresponding standalone tools",
				"  - Prefer `vp check` for validation loops",
				"",
				"  Finally, verify the migration by running:",
				`  - ${accent("vp install")}`,
				`  - ${accent("vp check")}`,
				`  - ${accent("vp test")}`,
				`  - ${accent("vp build")}`,
				"",
				"  Summarize the migration at the end and report any manual follow-up still",
				"  required."
			]
		}
	]
});
function parseArgs() {
	const parsed = lib_default(process.argv.slice(3), {
		alias: { h: "help" },
		boolean: [
			"help",
			"interactive",
			"hooks"
		],
		default: { interactive: defaultInteractive() }
	});
	const interactive = parsed.interactive;
	let projectPath = parsed._[0];
	if (projectPath) projectPath = path.resolve(process.cwd(), projectPath);
	else projectPath = process.cwd();
	return {
		projectPath,
		options: {
			interactive,
			help: parsed.help,
			agent: parsed.agent,
			editor: parsed.editor,
			hooks: parsed.hooks
		}
	};
}
async function collectMigrationPlan(rootDir, detectedPackageManager, options, packages) {
	const packageManager = detectedPackageManager ?? await selectPackageManager(options.interactive, true);
	let shouldSetupHooks = await promptGitHooks(options);
	if (shouldSetupHooks) {
		const reason = preflightGitHooksSetup(rootDir);
		if (reason) {
			log$1.warn(`⚠ ${reason}`);
			shouldSetupHooks = false;
		}
	}
	const existingAgentTargetPaths = options.agent !== void 0 || !options.interactive ? void 0 : detectExistingAgentTargetPaths(rootDir);
	const selectedAgentTargetPaths = existingAgentTargetPaths !== void 0 ? existingAgentTargetPaths : await selectAgentTargetPaths({
		interactive: options.interactive,
		agent: options.agent,
		onCancel: () => cancelAndExit()
	});
	const agentConflicts = await detectAgentConflicts({
		projectRoot: rootDir,
		targetPaths: selectedAgentTargetPaths
	});
	const agentConflictDecisions = /* @__PURE__ */ new Map();
	for (const conflict of agentConflicts) if (options.interactive) {
		const action = await select({
			message: `Agent instructions already exist at ${conflict.targetPath}.\n  ` + styleText("gray", "The Vite+ template includes guidance on `vp` commands, the build pipeline, and project conventions."),
			options: [{
				label: "Append",
				value: "append",
				hint: "Add template content to the end"
			}, {
				label: "Skip",
				value: "skip",
				hint: "Leave existing file unchanged"
			}],
			initialValue: "skip"
		});
		if (q(action)) cancelAndExit();
		agentConflictDecisions.set(conflict.targetPath, action);
	} else agentConflictDecisions.set(conflict.targetPath, "skip");
	const selectedEditor = await selectEditor({
		interactive: options.interactive,
		editor: options.editor,
		onCancel: () => cancelAndExit()
	});
	const editorConflicts = detectEditorConflicts({
		projectRoot: rootDir,
		editorId: selectedEditor
	});
	const editorConflictDecisions = /* @__PURE__ */ new Map();
	for (const conflict of editorConflicts) if (options.interactive) {
		const action = await select({
			message: `${conflict.displayPath} already exists.\n  ` + styleText("gray", "Vite+ adds editor settings for the built-in linter and formatter. Merge adds new keys without overwriting existing ones."),
			options: [{
				label: "Merge",
				value: "merge",
				hint: "Merge new settings into existing file"
			}, {
				label: "Skip",
				value: "skip",
				hint: "Leave existing file unchanged"
			}],
			initialValue: "skip"
		});
		if (q(action)) cancelAndExit();
		editorConflictDecisions.set(conflict.fileName, action);
	} else editorConflictDecisions.set(conflict.fileName, "merge");
	const eslintProject = detectEslintProject(rootDir, packages);
	let migrateEslint = false;
	if (eslintProject.hasDependency && !eslintProject.configFile && eslintProject.legacyConfigFile) warnLegacyEslintConfig(eslintProject.legacyConfigFile);
	else if (eslintProject.hasDependency && eslintProject.configFile) migrateEslint = await confirmEslintMigration(options.interactive);
	else if (eslintProject.hasDependency) warnPackageLevelEslint();
	const prettierProject = detectPrettierProject(rootDir, packages);
	let migratePrettier = false;
	if (prettierProject.hasDependency && prettierProject.configFile) migratePrettier = await confirmPrettierMigration(options.interactive);
	else if (prettierProject.hasDependency) warnPackageLevelPrettier();
	const nodeVersionDetection = detectNodeVersionManagerFile(rootDir);
	let migrateNodeVersionFile = false;
	if (nodeVersionDetection) migrateNodeVersionFile = await confirmNodeVersionFileMigration(options.interactive, nodeVersionDetection);
	const allDetectedFrameworks = new Set(detectFramework(rootDir));
	for (const pkg of packages ?? []) for (const framework of detectFramework(path.join(rootDir, pkg.path))) allDetectedFrameworks.add(framework);
	const frameworkShimFrameworks = [];
	for (const framework of allDetectedFrameworks) if (detectFramework(rootDir).includes(framework) && !hasFrameworkShim(rootDir, framework) || (packages ?? []).some((pkg) => {
		const pkgPath = path.join(rootDir, pkg.path);
		return detectFramework(pkgPath).includes(framework) && !hasFrameworkShim(pkgPath, framework);
	})) {
		if (await confirmFrameworkShim(framework, options.interactive)) frameworkShimFrameworks.push(framework);
	}
	return {
		packageManager,
		shouldSetupHooks,
		selectedAgentTargetPaths,
		agentConflictDecisions,
		selectedEditor,
		editorConflictDecisions,
		migrateEslint,
		eslintConfigFile: eslintProject.configFile,
		migratePrettier,
		prettierConfigFile: prettierProject.configFile,
		migrateNodeVersionFile,
		nodeVersionDetection,
		frameworkShimFrameworks: frameworkShimFrameworks.length > 0 ? frameworkShimFrameworks : void 0
	};
}
function formatDuration(durationMs) {
	if (durationMs < 1e3) return `${Math.max(1, durationMs)}ms`;
	const durationSeconds = durationMs / 1e3;
	if (durationSeconds < 10) return `${durationSeconds.toFixed(1)}s`;
	return `${Math.round(durationSeconds)}s`;
}
function showMigrationSummary(options) {
	const { projectRoot, packageManager, packageManagerVersion, installDurationMs, report, updatedExistingVitePlus } = options;
	const projectLabel = displayRelative(projectRoot) || ".";
	const configUpdates = report.createdViteConfigCount + report.mergedConfigCount + report.mergedStagedConfigCount + report.inlinedLintStagedConfigCount + report.removedConfigCount + report.tsdownImportCount;
	log(`${styleText("magenta", "◇")} ${updatedExistingVitePlus ? "Updated" : "Migrated"} ${accent(projectLabel)}${updatedExistingVitePlus ? "" : " to Vite+"}`);
	log(`${styleText("gray", "•")} Node ${process.versions.node}  ${packageManager} ${packageManagerVersion}`);
	if (installDurationMs > 0) log(`${styleText("green", "✓")} Dependencies installed in ${formatDuration(installDurationMs)}`);
	if (configUpdates > 0 || report.rewrittenImportFileCount > 0) {
		const parts = [];
		if (configUpdates > 0) parts.push(`${configUpdates} ${configUpdates === 1 ? "config update" : "config updates"} applied`);
		if (report.rewrittenImportFileCount > 0) parts.push(`${report.rewrittenImportFileCount} ${report.rewrittenImportFileCount === 1 ? "file had" : "files had"} imports rewritten`);
		log(`${styleText("gray", "•")} ${parts.join(", ")}`);
	}
	if (report.eslintMigrated) log(`${styleText("gray", "•")} ESLint rules migrated to Oxlint`);
	if (report.prettierMigrated) log(`${styleText("gray", "•")} Prettier migrated to Oxfmt`);
	if (report.nodeVersionFileMigrated) log(`${styleText("gray", "•")} Node version manager file migrated to .node-version`);
	if (report.gitHooksConfigured) log(`${styleText("gray", "•")} Git hooks configured`);
	if (report.frameworkShimAdded) log(`${styleText("gray", "•")} TypeScript shim added for framework component files`);
	if (report.warnings.length > 0) {
		log(`${styleText("yellow", "!")} Warnings:`);
		for (const warning of report.warnings) log(`  - ${warning}`);
	}
	if (report.manualSteps.length > 0) {
		log(`${styleText("blue", "→")} Manual follow-up:`);
		for (const step of report.manualSteps) log(`  - ${step}`);
	}
}
async function checkRolldownCompatibility(rootDir, report) {
	try {
		const { resolveConfig } = await import("../index.js");
		const { checkManualChunksCompat } = await import("../compat-DdC7fHjB.js");
		checkManualChunksCompat((await resolveConfig({
			root: rootDir,
			logLevel: "silent",
			configLoader: "runner"
		}, "build")).build?.rollupOptions?.output, report);
	} catch {}
}
async function executeMigrationPlan(workspaceInfoOptional, plan, interactive) {
	const report = createMigrationReport();
	const migrationProgress = interactive ? spinner({ indicator: "timer" }) : void 0;
	let migrationProgressStarted = false;
	const updateMigrationProgress = (message) => {
		if (!migrationProgress) return;
		if (migrationProgressStarted) {
			migrationProgress.message(message);
			return;
		}
		migrationProgress.start(message);
		migrationProgressStarted = true;
	};
	const clearMigrationProgress = () => {
		if (migrationProgress && migrationProgressStarted) {
			migrationProgress.clear();
			migrationProgressStarted = false;
		}
	};
	const failMigrationProgress = (message) => {
		if (migrationProgress && migrationProgressStarted) {
			migrationProgress.error(message);
			migrationProgressStarted = false;
		}
	};
	updateMigrationProgress("Preparing migration");
	const downloadResult = await downloadPackageManager(plan.packageManager, workspaceInfoOptional.packageManagerVersion, interactive, true);
	const workspaceInfo = {
		...workspaceInfoOptional,
		packageManager: plan.packageManager,
		downloadPackageManager: downloadResult
	};
	if (plan.packageManager === PackageManager.yarn && import_semver.default.satisfies(downloadResult.version, ">=4.0.0 <4.10.0")) {
		updateMigrationProgress("Upgrading Yarn");
		await upgradeYarn(workspaceInfo.rootDir, interactive, true);
	} else if (plan.packageManager === PackageManager.pnpm && import_semver.default.satisfies(downloadResult.version, "< 9.5.0")) {
		failMigrationProgress("Migration failed");
		log$1.error(`✘ pnpm@${downloadResult.version} is not supported by auto migration, please upgrade pnpm to >=9.5.0 first`);
		cancelAndExit("Vite+ cannot automatically migrate this project yet.", 1);
	} else if (plan.packageManager === PackageManager.npm && import_semver.default.satisfies(downloadResult.version, "< 8.3.0")) {
		failMigrationProgress("Migration failed");
		log$1.error(`✘ npm@${downloadResult.version} is not supported by auto migration, please upgrade npm to >=8.3.0 first`);
		cancelAndExit("Vite+ cannot automatically migrate this project yet.", 1);
	}
	if (plan.migrateNodeVersionFile && plan.nodeVersionDetection) {
		updateMigrationProgress("Migrating node version file");
		migrateNodeVersionManagerFile(workspaceInfo.rootDir, plan.nodeVersionDetection, report);
	}
	updateMigrationProgress("Installing dependencies");
	const initialInstallSummary = await runViteInstall(workspaceInfo.rootDir, interactive, void 0, { silent: true });
	updateMigrationProgress("Validating toolchain");
	const isViteSupported = checkViteVersion(workspaceInfo.rootDir);
	const isVitestSupported = checkVitestVersion(workspaceInfo.rootDir);
	if (!isViteSupported || !isVitestSupported) {
		failMigrationProgress("Migration failed");
		cancelAndExit("Vite+ cannot automatically migrate this project yet.", 1);
	}
	updateMigrationProgress("Checking config compatibility");
	await checkRolldownCompatibility(workspaceInfo.rootDir, report);
	if (workspaceInfo.packages) for (const pkg of workspaceInfo.packages) await checkRolldownCompatibility(path.join(workspaceInfo.rootDir, pkg.path), report);
	if (plan.migrateEslint) {
		updateMigrationProgress("Migrating ESLint");
		if (!await migrateEslintToOxlint(workspaceInfo.rootDir, interactive, plan.eslintConfigFile, workspaceInfo.packages, {
			silent: true,
			report
		})) {
			failMigrationProgress("Migration failed");
			cancelAndExit("ESLint migration failed. Fix the issue and re-run `vp migrate`.", 1);
		}
	}
	if (plan.migratePrettier) {
		updateMigrationProgress("Migrating Prettier");
		if (!await migratePrettierToOxfmt(workspaceInfo.rootDir, interactive, plan.prettierConfigFile, workspaceInfo.packages, {
			silent: true,
			report
		})) {
			failMigrationProgress("Migration failed");
			cancelAndExit("Prettier migration failed. Fix the issue and re-run `vp migrate`.", 1);
		}
	}
	const skipStagedMigration = !plan.shouldSetupHooks;
	updateMigrationProgress("Rewriting configs");
	if (workspaceInfo.isMonorepo) rewriteMonorepo(workspaceInfo, skipStagedMigration, true, report);
	else rewriteStandaloneProject(workspaceInfo.rootDir, workspaceInfo, skipStagedMigration, true, report);
	if (plan.shouldSetupHooks) {
		updateMigrationProgress("Configuring git hooks");
		installGitHooks(workspaceInfo.rootDir, true, report);
	}
	updateMigrationProgress("Writing agent instructions");
	await writeAgentInstructions({
		projectRoot: workspaceInfo.rootDir,
		targetPaths: plan.selectedAgentTargetPaths,
		interactive,
		conflictDecisions: plan.agentConflictDecisions,
		silent: true
	});
	updateMigrationProgress("Writing editor configs");
	await writeEditorConfigs({
		projectRoot: workspaceInfo.rootDir,
		editorId: plan.selectedEditor,
		interactive,
		conflictDecisions: plan.editorConflictDecisions,
		silent: true
	});
	if (plan.frameworkShimFrameworks) {
		updateMigrationProgress("Adding TypeScript shim");
		for (const framework of plan.frameworkShimFrameworks) {
			if (detectFramework(workspaceInfo.rootDir).includes(framework) && !hasFrameworkShim(workspaceInfo.rootDir, framework)) addFrameworkShim(workspaceInfo.rootDir, framework, report);
			for (const pkg of workspaceInfo.packages) {
				const pkgPath = path.join(workspaceInfo.rootDir, pkg.path);
				if (detectFramework(pkgPath).includes(framework) && !hasFrameworkShim(pkgPath, framework)) addFrameworkShim(pkgPath, framework, report);
			}
		}
	}
	const installArgs = plan.packageManager === PackageManager.npm || plan.packageManager === PackageManager.bun ? ["--force"] : void 0;
	updateMigrationProgress("Installing dependencies");
	const finalInstallSummary = await runViteInstall(workspaceInfo.rootDir, interactive, installArgs, { silent: true });
	clearMigrationProgress();
	return {
		installDurationMs: initialInstallSummary.durationMs + finalInstallSummary.durationMs,
		packageManagerVersion: downloadResult.version,
		report
	};
}
async function main() {
	const { projectPath, options } = parseArgs();
	if (options.help) {
		printHeader();
		log(helpMessage);
		return;
	}
	printHeader();
	const workspaceInfoOptional = await detectWorkspace(projectPath);
	const resolvedPackageManager = workspaceInfoOptional.packageManager ?? "unknown";
	const rootPkg = readNearestPackageJson(workspaceInfoOptional.rootDir);
	if (hasVitePlusDependency(rootPkg) && !isForceOverrideMode()) {
		let didMigrate = false;
		let installDurationMs = 0;
		const report = createMigrationReport();
		const migrationProgress = options.interactive ? spinner({ indicator: "timer" }) : void 0;
		let migrationProgressStarted = false;
		const updateMigrationProgress = (message) => {
			if (!migrationProgress) return;
			if (migrationProgressStarted) {
				migrationProgress.message(message);
				return;
			}
			migrationProgress.start(message);
			migrationProgressStarted = true;
		};
		const clearMigrationProgress = () => {
			if (migrationProgress && migrationProgressStarted) {
				migrationProgress.clear();
				migrationProgressStarted = false;
			}
		};
		const eslintMigrated = await promptEslintMigration(workspaceInfoOptional.rootDir, options.interactive, workspaceInfoOptional.packages);
		const prettierMigrated = await promptPrettierMigration(workspaceInfoOptional.rootDir, options.interactive, workspaceInfoOptional.packages);
		const nodeVersionDetection = detectNodeVersionManagerFile(workspaceInfoOptional.rootDir);
		if (nodeVersionDetection) {
			if (await confirmNodeVersionFileMigration(options.interactive, nodeVersionDetection) && migrateNodeVersionManagerFile(workspaceInfoOptional.rootDir, nodeVersionDetection, report)) didMigrate = true;
		}
		if (eslintMigrated || prettierMigrated) {
			updateMigrationProgress("Rewriting configs");
			mergeViteConfigFiles(workspaceInfoOptional.rootDir, true, report);
			updateMigrationProgress("Installing dependencies");
			const installSummary = await runViteInstall(workspaceInfoOptional.rootDir, options.interactive, void 0, { silent: true });
			installDurationMs += installSummary.durationMs;
			didMigrate = true;
			report.eslintMigrated = eslintMigrated;
			report.prettierMigrated = prettierMigrated;
		}
		if (rootPkg?.devDependencies?.husky || rootPkg?.dependencies?.husky || rootPkg?.devDependencies?.["lint-staged"] || rootPkg?.dependencies?.["lint-staged"]) {
			const shouldSetupHooks = await promptGitHooks(options);
			if (shouldSetupHooks) updateMigrationProgress("Configuring git hooks");
			if (shouldSetupHooks && installGitHooks(workspaceInfoOptional.rootDir, true, report)) didMigrate = true;
		}
		await checkRolldownCompatibility(workspaceInfoOptional.rootDir, report);
		if (workspaceInfoOptional.packages) for (const pkg of workspaceInfoOptional.packages) await checkRolldownCompatibility(path.join(workspaceInfoOptional.rootDir, pkg.path), report);
		if (didMigrate || report.warnings.length > 0) {
			clearMigrationProgress();
			showMigrationSummary({
				projectRoot: workspaceInfoOptional.rootDir,
				packageManager: resolvedPackageManager,
				packageManagerVersion: workspaceInfoOptional.packageManagerVersion,
				installDurationMs,
				report,
				updatedExistingVitePlus: true
			});
		} else outro(`This project is already using Vite+! ${accent(`Happy coding!`)}`);
		return;
	}
	const plan = await collectMigrationPlan(workspaceInfoOptional.rootDir, workspaceInfoOptional.packageManager, options, workspaceInfoOptional.packages);
	const result = await executeMigrationPlan(workspaceInfoOptional, plan, options.interactive);
	showMigrationSummary({
		projectRoot: workspaceInfoOptional.rootDir,
		packageManager: plan.packageManager,
		packageManagerVersion: result.packageManagerVersion,
		installDurationMs: result.installDurationMs,
		report: result.report
	});
}
main().catch((err) => {
	log$1.error(err.message);
	console.error(err);
	process.exit(1);
});
//#endregion
export {};
