// vite-plus@0.1.20/dist/bin.js
import { a as DEFAULT_ENVS, i as BASEURL_TSCONFIG_WARNING, o as VITE_PLUS_NAME, u as resolve$1 } from "./main-A6UrSTYb.js";
import { i as runCommandSilently, n as hasBaseUrlInTsconfig } from "./tsconfig-DQTf06oN.js";
import { n as errorMsg, r as log, s as warnMsg, t as accent } from "./terminal-CxTMfsxZ.js";
import { t as resolveUniversalViteConfig } from "./resolve-vite-config-B_w1u1j4.js";
import path, { dirname, join } from "node:path";
import { mergeJsonConfig, run } from "../binding/index.js";
import fs, { existsSync, realpathSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { relative } from "node:path/win32";
//#region src/resolve-fmt.ts
/**
* Oxfmt tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the oxfmt binary path
* using Node.js module resolution. The resolved path is passed back
* to the Rust core, which then executes oxfmt for code formatting.
*
* Used for: `vite-plus fmt` command
*
* Oxfmt is a fast JavaScript/TypeScript formatter written in Rust that
* provides high-performance code formatting capabilities.
*/
/**
* Resolves the oxfmt binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the oxfmt binary
*   - envs: Environment variables to set when executing oxfmt
*
* The environment variables provide runtime context to oxfmt,
* including Node.js version information and package manager details.
*/
async function fmt() {
	return {
		binPath: join(dirname(dirname(resolve$1("oxfmt"))), "bin", "oxfmt"),
		envs: { ...DEFAULT_ENVS }
	};
}
//#endregion
//#region src/init-config.ts
const INIT_COMMAND_SPECS = {
	lint: {
		configKey: "lint",
		triggerFlags: ["--init"],
		defaultConfigFiles: [".oxlintrc.json"]
	},
	fmt: {
		configKey: "fmt",
		triggerFlags: ["--init", "--migrate"],
		defaultConfigFiles: [".oxfmtrc.json", ".oxfmtrc.jsonc"]
	}
};
const VITE_CONFIG_FILES = [
	"vite.config.ts",
	"vite.config.mts",
	"vite.config.cts",
	"vite.config.js",
	"vite.config.mjs",
	"vite.config.cjs"
];
function optionTerminatorIndex(args) {
	const index = args.indexOf("--");
	return index === -1 ? args.length : index;
}
function hasTriggerFlag(args, triggerFlags) {
	const limit = optionTerminatorIndex(args);
	for (let i = 0; i < limit; i++) {
		const arg = args[i];
		if (triggerFlags.some((flag) => arg === flag || arg.startsWith(`${flag}=`))) return true;
	}
	return false;
}
function extractConfigPathArg(args) {
	const limit = optionTerminatorIndex(args);
	for (let i = 0; i < limit; i++) {
		const arg = args[i];
		if (arg === "-c" || arg === "--config") {
			const value = args[i + 1];
			return value ? value : null;
		}
		if (arg.startsWith("--config=")) return arg.slice(9);
		if (arg.startsWith("-c=")) return arg.slice(3);
	}
	return null;
}
function resolveGeneratedConfigPath(projectPath, args, defaultConfigFiles) {
	const configArg = extractConfigPathArg(args);
	if (configArg) {
		const resolved = path.isAbsolute(configArg) ? configArg : path.join(projectPath, configArg);
		if (fs.existsSync(resolved)) return resolved;
	}
	for (const filename of defaultConfigFiles) {
		const fullPath = path.join(projectPath, filename);
		if (fs.existsSync(fullPath)) return fullPath;
	}
	return null;
}
function findViteConfigPath(projectPath) {
	for (const filename of VITE_CONFIG_FILES) {
		const fullPath = path.join(projectPath, filename);
		if (fs.existsSync(fullPath)) return fullPath;
	}
	return null;
}
function ensureViteConfigPath(projectPath) {
	const existing = findViteConfigPath(projectPath);
	if (existing) return existing;
	const viteConfigPath = path.join(projectPath, "vite.config.ts");
	fs.writeFileSync(viteConfigPath, `import { defineConfig } from '${VITE_PLUS_NAME}';

export default defineConfig({});
`);
	return viteConfigPath;
}
function hasConfigKey(viteConfigPath, configKey) {
	const viteConfig = fs.readFileSync(viteConfigPath, "utf8");
	return new RegExp(`\\b${configKey}\\s*:`).test(viteConfig);
}
async function vpFmt(cwd, filePath) {
	const { binPath, envs } = await fmt();
	const result = await runCommandSilently({
		command: binPath,
		args: ["--write", filePath],
		cwd,
		envs: {
			...process.env,
			...envs
		}
	});
	if (result.exitCode !== 0) warnMsg(`Failed to format ${filePath} with vp fmt:\n${result.stdout.toString()}${result.stderr.toString()}`);
}
function resolveInitSpec(command, args) {
	if (!command) return null;
	const spec = INIT_COMMAND_SPECS[command];
	if (!spec || !hasTriggerFlag(args, spec.triggerFlags)) return null;
	return spec;
}
function inspectInitCommand(command, args, projectPath = process.cwd()) {
	const spec = resolveInitSpec(command, args);
	if (!spec) return { handled: false };
	const viteConfigPath = findViteConfigPath(projectPath);
	if (!viteConfigPath) return {
		handled: true,
		configKey: spec.configKey,
		hasExistingConfigKey: false
	};
	return {
		handled: true,
		configKey: spec.configKey,
		existingViteConfigPath: viteConfigPath,
		hasExistingConfigKey: hasConfigKey(viteConfigPath, spec.configKey)
	};
}
/**
* Merge generated tool config from `vp lint/fmt --init` (and fmt --migrate)
* into the project's vite config, then remove the generated standalone file.
*
* Returns true when the command was an init/migrate command (handled), false otherwise.
*/
async function applyToolInitConfigToViteConfig(command, args, projectPath = process.cwd()) {
	const inspection = inspectInitCommand(command, args, projectPath);
	if (!inspection.handled || !inspection.configKey) return { handled: false };
	const spec = INIT_COMMAND_SPECS[command];
	const viteConfigPath = ensureViteConfigPath(projectPath);
	const generatedConfigPath = resolveGeneratedConfigPath(projectPath, args, spec.defaultConfigFiles);
	if (hasConfigKey(viteConfigPath, spec.configKey)) {
		if (generatedConfigPath) fs.rmSync(generatedConfigPath, { force: true });
		return {
			handled: true,
			action: "skipped-existing",
			configKey: spec.configKey,
			viteConfigPath
		};
	}
	if (spec.configKey === "lint" && hasTriggerFlag(args, ["--init"])) {
		const lintInitConfigPath = path.join(projectPath, ".vite-plus-lint-init.oxlintrc.json");
		const hasBaseUrl = hasBaseUrlInTsconfig(projectPath);
		const initOptions = hasBaseUrl ? {} : {
			typeAware: true,
			typeCheck: true
		};
		if (hasBaseUrl) warnMsg(BASEURL_TSCONFIG_WARNING);
		fs.writeFileSync(lintInitConfigPath, JSON.stringify({ options: initOptions }));
		const mergeResult = mergeJsonConfig(viteConfigPath, lintInitConfigPath, spec.configKey);
		if (!mergeResult.updated) throw new Error(`Failed to initialize lint config in ${path.basename(viteConfigPath)}`);
		fs.writeFileSync(viteConfigPath, mergeResult.content);
		fs.rmSync(lintInitConfigPath, { force: true });
		if (generatedConfigPath) fs.rmSync(generatedConfigPath, { force: true });
		await vpFmt(projectPath, path.relative(projectPath, viteConfigPath));
		return {
			handled: true,
			action: "added",
			configKey: spec.configKey,
			viteConfigPath
		};
	}
	if (!generatedConfigPath) return {
		handled: true,
		action: "no-generated-config",
		configKey: inspection.configKey,
		viteConfigPath
	};
	const mergeResult = mergeJsonConfig(viteConfigPath, generatedConfigPath, spec.configKey);
	if (!mergeResult.updated) throw new Error(`Failed to merge ${path.basename(generatedConfigPath)} into ${path.basename(viteConfigPath)}`);
	fs.writeFileSync(viteConfigPath, mergeResult.content);
	fs.rmSync(generatedConfigPath, { force: true });
	await vpFmt(projectPath, path.relative(projectPath, viteConfigPath));
	return {
		handled: true,
		action: "added",
		configKey: spec.configKey,
		viteConfigPath
	};
}
//#endregion
//#region src/resolve-doc.ts
/**
* VitePress tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the VitePress binary path
* to the bundled VitePress in the CLI distribution. The resolved path is
* passed back to the Rust core, which then executes VitePress with the
* appropriate command and arguments.
*
* Used for: `vite doc` command
*/
/**
* Resolves the VitePress binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the VitePress CLI entry point (vitepress.js)
*   - envs: Environment variables to set when executing VitePress
*
* The function points to the bundled VitePress in the CLI's dist directory.
*/
async function doc() {
	return {
		binPath: join(dirname(fileURLToPath(import.meta.url)), "vitepress", "node", "cli.js"),
		envs: { ...DEFAULT_ENVS }
	};
}
//#endregion
//#region src/resolve-lint.ts
/**
* Oxlint tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the oxlint binary path
* using Node.js module resolution. The resolved path is passed back
* to the Rust core, which then executes oxlint for code linting.
*
* Used for: `vite-plus lint` command
*
* Oxlint is a fast JavaScript/TypeScript linter written in Rust that
* provides ESLint-compatible linting with significantly better performance.
*/
/**
* Resolves the oxlint binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the oxlint binary
*   - envs: Environment variables to set when executing oxlint
*
* The environment variables provide runtime context to oxlint,
* including Node.js version information and package manager details.
*/
async function lint() {
	const binPath = join(dirname(dirname(resolve$1("oxlint"))), "bin", "oxlint");
	let oxlintTsgolintPath = resolve$1("oxlint-tsgolint/bin/tsgolint");
	if (process.platform === "win32") {
		const scriptDir = dirname(fileURLToPath(import.meta.url));
		const localBinDir = join(scriptDir, "..", "node_modules", ".bin");
		const projectBinDir = join(dirname(dirname(oxlintTsgolintPath)), "..", ".bin");
		const pathCandidates = [
			join(localBinDir, "tsgolint.exe"),
			join(localBinDir, "tsgolint.cmd"),
			join(projectBinDir, "tsgolint.exe"),
			join(projectBinDir, "tsgolint.cmd")
		];
		oxlintTsgolintPath = pathCandidates.find((p) => existsSync(p)) ?? "";
		if (!oxlintTsgolintPath) try {
			const realBinDir = join(dirname(realpathSync(join(scriptDir, ".."))), ".bin");
			oxlintTsgolintPath = [join(realBinDir, "tsgolint.exe"), join(realBinDir, "tsgolint.cmd")].find((p) => existsSync(p)) ?? "";
		} catch {}
		if (!oxlintTsgolintPath) throw new Error("Unable to resolve oxlint-tsgolint executable, tried:\n" + pathCandidates.map((path) => `- ${path}`).join("\n"));
		const relativePath = relative(process.cwd(), oxlintTsgolintPath);
		oxlintTsgolintPath = /^[a-zA-Z]:/.test(relativePath) ? relativePath : `.\\${relativePath}`;
	}
	return {
		binPath,
		envs: {
			...DEFAULT_ENVS,
			OXLINT_TSGOLINT_PATH: oxlintTsgolintPath
		}
	};
}
//#endregion
//#region src/resolve-pack.ts
/**
* Tsdown tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the Tsdown binary path
* using Node.js module resolution. The resolved path is passed back
* to the Rust core, which then executes Tsdown for running pack.
*
* Used for: `vite-plus pack` command
*/
/**
* Resolves the Tsdown binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the Tsdown CLI entry point
*   - envs: Environment variables to set when executing Tsdown
*
* Tsdown is a tool that provides a library for building JavaScript/TypeScript libraries.
*/
async function pack() {
	return {
		binPath: join(import.meta.dirname, "pack-bin.js"),
		envs: { ...DEFAULT_ENVS }
	};
}
//#endregion
//#region src/resolve-test.ts
/**
* Vitest tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the Vitest binary path
* to the bundled Vitest in the CLI distribution. The resolved path is
* passed back to the Rust core, which then executes Vitest for running tests.
*
* Used for: `vite-plus test` command
*/
/**
* Resolves the Vitest binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the Vitest CLI entry point (vitest.mjs)
*   - envs: Environment variables to set when executing Vitest
*
* Vitest is Vite's testing framework that provides a Jest-compatible
* testing experience with Vite's fast HMR and transformation pipeline.
* The function points to the bundled Vitest in the CLI's dist directory.
*/
async function test() {
	return {
		binPath: join(dirname(resolve$1("@voidzero-dev/vite-plus-test")), "dist", "cli.js"),
		envs: process.env.DEBUG_DISABLE_SOURCE_MAP ? {
			...DEFAULT_ENVS,
			DEBUG_DISABLE_SOURCE_MAP: process.env.DEBUG_DISABLE_SOURCE_MAP
		} : { ...DEFAULT_ENVS }
	};
}
//#endregion
//#region src/resolve-vite.ts
/**
* Vite tool resolver for the vite-plus CLI.
*
* This module exports a function that resolves the Vite binary path
* using Node.js module resolution. The resolved path is passed back
* to the Rust core, which then executes Vite with the appropriate
* command and arguments.
*
* Used for: `vite-plus build` and potentially `vite-plus dev` commands
*/
/**
* Resolves the Vite binary path and environment variables.
*
* @returns Promise containing:
*   - binPath: Absolute path to the Vite CLI entry point (vite.js)
*   - envs: Environment variables to set when executing Vite
*
* The function first tries to resolve vite package, then falls back
* to vite package (for direct vite installations).
* It constructs the path to the CLI binary within the resolved package.
*/
async function vite() {
	return {
		binPath: join(dirname(resolve$1("@voidzero-dev/vite-plus-core")), "cli.js"),
		envs: process.env.DEBUG_DISABLE_SOURCE_MAP ? {
			...DEFAULT_ENVS,
			DEBUG_DISABLE_SOURCE_MAP: process.env.DEBUG_DISABLE_SOURCE_MAP
		} : { ...DEFAULT_ENVS }
	};
}
//#endregion
//#region src/bin.ts
/**
* Unified entry point for both the local CLI (via bin/vp) and the global CLI (via Rust vp binary).
*
* Global commands (create, migrate, config, staged, --version) are handled by tsdown-bundled modules.
* All other commands are delegated to the Rust core through NAPI bindings, which
* uses JavaScript tool resolver functions to locate tool binaries.
*
* When called from the global CLI, the Rust binary resolves the project's local
* vite-plus installation using oxc_resolver and runs its dist/bin.js directly.
* If no local installation is found, this global dist/bin.js is used as fallback.
*/
function getErrorMessage(err) {
	if (err instanceof Error) return err.message;
	if (typeof err === "object" && err && "message" in err && typeof err.message === "string") return err.message;
	return String(err);
}
let args = process.argv.slice(2);
if (args[0] === "help" && args[1]) {
	args = [
		args[1],
		"--help",
		...args.slice(2)
	];
	process.argv = process.argv.slice(0, 2).concat(args);
}
const command = args[0];
process.env.VP_COMMAND = command ?? "";
if (command === "create") await import("./create/bin.js");
else if (command === "migrate") await import("./migration/bin.js");
else if (command === "config") await import("./config/bin.js");
else if (command === "--version" || command === "-V") await import("./version.js");
else if (command === "staged") await import("./staged/bin.js");
else try {
	const initInspection = inspectInitCommand(command, args.slice(1));
	if (initInspection.handled && initInspection.configKey && initInspection.hasExistingConfigKey && initInspection.existingViteConfigPath) {
		log(`Skipped initialization: '${accent(initInspection.configKey)}' already exists in '${accent(path.basename(initInspection.existingViteConfigPath))}'.`);
		process.exit(0);
	}
	const exitCode = await run({
		lint,
		pack,
		fmt,
		vite,
		test,
		doc,
		resolveUniversalViteConfig,
		args: process.argv.slice(2)
	});
	let finalExitCode = exitCode;
	if (exitCode === 0) try {
		const result = await applyToolInitConfigToViteConfig(command, args.slice(1));
		if (result.handled && result.action === "added" && result.configKey && result.viteConfigPath) log(`Added '${accent(result.configKey)}' to '${accent(path.basename(result.viteConfigPath))}'.`);
		if (result.handled && result.action === "skipped-existing" && result.configKey && result.viteConfigPath) log(`Skipped initialization: '${accent(result.configKey)}' already exists in '${accent(path.basename(result.viteConfigPath))}'.`);
	} catch (err) {
		console.error("[Vite+] Failed to initialize config in vite.config.ts:", err);
		finalExitCode = 1;
	}
	process.exit(finalExitCode);
} catch (err) {
	errorMsg(getErrorMessage(err));
	process.exit(1);
}
//#endregion
export {};
