// vite-plus@0.1.20/dist/version.js
import "./main-A6UrSTYb.js";
import { a as printHeader, r as log, t as accent } from "./terminal-CxTMfsxZ.js";
import { i as hasVitePlusDependency, n as detectPackageMetadata } from "./package-D_LD1iiI.js";
import { t as renderCliDoc } from "./help-BtkjXtRM.js";
import path from "node:path";
import fs from "node:fs";
//#region package.json
var version = "0.0.0";
//#endregion
//#region src/version.ts
/** Tool display names in the order shown by `vp --version`. */
const TOOL_DISPLAY_ORDER = [
	"vite",
	"rolldown",
	"vitest",
	"oxfmt",
	"oxlint",
	"oxlint-tsgolint",
	"tsdown"
];
function getGlobalVersion() {
	return process.env.VP_GLOBAL_VERSION ?? null;
}
function getCliVersion() {
	return version ?? null;
}
function getLocalMetadata(cwd) {
	if (!isVitePlusDeclaredInAncestors(cwd)) return null;
	return detectPackageMetadata(cwd, "vite-plus") ?? null;
}
function isVitePlusDeclaredInAncestors(cwd) {
	let currentDir = path.resolve(cwd);
	while (true) {
		const packageJsonPath = path.join(currentDir, "package.json");
		try {
			if (hasVitePlusDependency(JSON.parse(fs.readFileSync(packageJsonPath, "utf8")))) return true;
		} catch {}
		const parentDir = path.dirname(currentDir);
		if (parentDir === currentDir) break;
		currentDir = parentDir;
	}
	return false;
}
/**
* Resolve all tool versions from the locally installed vite-plus package.
* Uses the `vite-plus/versions` export generated by `syncVersionsExport()`.
*/
async function resolveToolVersions(localPackagePath) {
	try {
		const mod = await import(`${localPackagePath}/dist/versions.js`);
		if (mod.versions && typeof mod.versions === "object") return mod.versions;
	} catch {}
	return {};
}
/**
* Print version information
*/
async function printVersion(cwd) {
	const globalVersion = getGlobalVersion();
	const cliVersion = getCliVersion();
	const localMetadata = getLocalMetadata(cwd);
	const localVersion = localMetadata?.version ?? null;
	const vpVersion = globalVersion ?? cliVersion ?? localVersion ?? "unknown";
	printHeader();
	log(`vp v${vpVersion}\n`);
	const sections = [{
		title: "Local vite-plus",
		rows: [{
			label: accent("vite-plus"),
			description: localVersion ? `v${localVersion}` : "Not found"
		}]
	}];
	if (localMetadata) {
		const versions = await resolveToolVersions(localMetadata.path);
		sections.push({
			title: "Tools",
			rows: TOOL_DISPLAY_ORDER.map((name) => ({
				label: accent(name),
				description: versions[name] ? `v${versions[name]}` : "Not found"
			}))
		});
	}
	log(renderCliDoc({ sections }));
}
await printVersion(process.cwd());
//#endregion
export { printVersion };
