// storybook@10.2.19/dist/babel/index.js
import CJS_COMPAT_NODE_URL_q3hucshmk7 from 'node:url';
import CJS_COMPAT_NODE_PATH_q3hucshmk7 from 'node:path';
import CJS_COMPAT_NODE_MODULE_q3hucshmk7 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_q3hucshmk7.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_q3hucshmk7.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_q3hucshmk7.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  BabelFileClass,
  babelParse,
  babelParseExpression,
  babelPrint,
  core,
  generate,
  import_core,
  parser,
  parserOptions,
  recast,
  traverse,
  types
} from "../_node-chunks/chunk-OOCT2CCT.js";
import "../_node-chunks/chunk-R37DSTA2.js";
import "../_node-chunks/chunk-ADTXI3R5.js";
import "../_node-chunks/chunk-GD5QW2LP.js";
var export_transformSync = import_core.transformSync;
export {
  BabelFileClass,
  babelParse,
  babelParseExpression,
  babelPrint,
  core,
  generate,
  parser,
  parserOptions,
  recast,
  export_transformSync as transformSync,
  traverse,
  types
};
