// vite-plus@0.1.20/dist/config/bin.js
import { a as printHeader, r as log } from "../terminal-CxTMfsxZ.js";
import { I as defaultInteractive, R as promptGitHooks, _ as hasStagedConfigInViteConfig, h as ensurePreCommitHook, i as updateExistingAgentInstructions } from "../agent-D_WSpD0r.js";
import { t as lib_default } from "../lib-BamM40b7.js";
import { t as renderCliDoc } from "../help-BtkjXtRM.js";
import { join } from "node:path";
import { existsSync, mkdirSync, rmSync, writeFileSync } from "node:fs";
import { spawnSync } from "node:child_process";
//#region src/config/hooks.ts
const HOOKS = [
	"pre-commit",
	"pre-merge-commit",
	"prepare-commit-msg",
	"commit-msg",
	"post-commit",
	"applypatch-msg",
	"pre-applypatch",
	"post-applypatch",
	"pre-rebase",
	"post-rewrite",
	"post-checkout",
	"post-merge",
	"pre-push",
	"pre-auto-gc"
];
function nestedDirname(depth) {
	let expr = "\"$0\"";
	for (let i = 0; i < depth; i++) expr = `"$(dirname ${expr})"`;
	return expr;
}
function hookScript(dir) {
	return `#!/usr/bin/env sh
{ [ "$HUSKY" = "2" ] || [ "$VITE_GIT_HOOKS" = "2" ]; } && set -x
n=$(basename "$0")
s=$(dirname "$(dirname "$0")")/$n

[ ! -f "$s" ] && exit 0

i="\${XDG_CONFIG_HOME:-$HOME/.config}/vite-plus/hooks-init.sh"
[ ! -f "$i" ] && i="\${XDG_CONFIG_HOME:-$HOME/.config}/husky/init.sh"
[ -f "$i" ] && . "$i"

{ [ "\${HUSKY-}" = "0" ] || [ "\${VITE_GIT_HOOKS-}" = "0" ]; } && exit 0

d=${nestedDirname(dir.split("/").filter((s) => s !== "" && s !== ".").length + 2)}
export PATH="$d/node_modules/.bin:$PATH"
sh -e "$s" "$@"
c=$?

[ $c != 0 ] && echo "VITE+ - $n script failed (code $c)"
[ $c = 127 ] && echo "VITE+ - command not found in PATH=$PATH"
exit $c`;
}
function install(dir = ".vite-hooks") {
	if (process.env.HUSKY === "0" || process.env.VITE_GIT_HOOKS === "0") return {
		message: "skip install (git hooks disabled)",
		isError: false
	};
	if (dir.includes("..")) return {
		message: ".. not allowed",
		isError: false
	};
	const prefixResult = spawnSync("git", ["rev-parse", "--show-prefix"]);
	if (prefixResult.status == null) return {
		message: "git command not found",
		isError: true
	};
	if (prefixResult.status !== 0) return {
		message: ".git can't be found",
		isError: false
	};
	const internal = (x = "") => join(dir, "_", x);
	const rel = prefixResult.stdout.toString().trim().replace(/\/$/, "");
	const target = rel ? `${rel}/${dir}/_` : `${dir}/_`;
	const checkResult = spawnSync("git", [
		"config",
		"--local",
		"core.hooksPath"
	]);
	const existingHooksPath = checkResult.status === 0 ? checkResult.stdout?.toString().trim() : "";
	if (existingHooksPath && existingHooksPath !== target && existingHooksPath !== ".husky" && !existingHooksPath.startsWith(".husky/")) return {
		message: `core.hooksPath is already set to "${existingHooksPath}", skipping`,
		isError: false
	};
	const { status, stderr } = spawnSync("git", [
		"config",
		"core.hooksPath",
		target
	]);
	if (status == null) return {
		message: "git command not found",
		isError: true
	};
	if (status) return {
		message: "" + stderr,
		isError: true
	};
	rmSync(internal("husky.sh"), { force: true });
	mkdirSync(internal(), { recursive: true });
	writeFileSync(internal(".gitignore"), "*");
	writeFileSync(internal("h"), hookScript(dir), { mode: 493 });
	for (const hook of HOOKS) writeFileSync(internal(hook), `#!/usr/bin/env sh\n. "$(dirname "$0")/h"`, { mode: 493 });
	return {
		message: "",
		isError: false
	};
}
//#endregion
//#region src/config/bin.ts
async function main() {
	const args = lib_default(process.argv.slice(3), {
		boolean: ["help", "hooks-only"],
		string: ["hooks-dir"],
		alias: { h: "help" }
	});
	if (args.help) {
		const helpMessage = renderCliDoc({
			usage: "vp config [OPTIONS]",
			summary: "Configure Vite+ for the current project (hooks + agent integration).",
			documentationUrl: "https://viteplus.dev/guide/commit-hooks",
			sections: [{
				title: "Options",
				rows: [{
					label: "--hooks-dir <path>",
					description: "Custom hooks directory (default: .vite-hooks)"
				}, {
					label: "-h, --help",
					description: "Show this help message"
				}]
			}, {
				title: "Environment",
				rows: [{
					label: "VITE_GIT_HOOKS=0",
					description: "Skip hook installation"
				}]
			}]
		});
		printHeader();
		log(helpMessage);
		return;
	}
	const dir = args["hooks-dir"];
	const hooksOnly = args["hooks-only"];
	const interactive = defaultInteractive();
	const lifecycleEvent = process.env.npm_lifecycle_event;
	const isLifecycleScript = lifecycleEvent === "prepare" || lifecycleEvent === "postinstall";
	const root = process.cwd();
	const hooksDir = dir ?? ".vite-hooks";
	const isFirstHooksRun = !existsSync(join(root, hooksDir, "_", "pre-commit"));
	let shouldSetupHooks = true;
	if (interactive && isFirstHooksRun && !dir && !isLifecycleScript && !hasStagedConfigInViteConfig(root)) shouldSetupHooks = await promptGitHooks({ interactive });
	if (shouldSetupHooks) {
		const { message, isError } = install(dir);
		if (message) {
			log(message);
			if (isError) process.exit(1);
		}
		if (!message) ensurePreCommitHook(root, hooksDir);
	}
	if (!hooksOnly) updateExistingAgentInstructions(root);
}
main();
//#endregion
export {};
