// @voidzero-dev/vite-plus-test@0.1.20/dist/reporters.js
export { M as AgentReporter, B as BenchmarkReporter, a as BenchmarkReportsMap, D as DefaultReporter, b as DotReporter, G as GithubActionsReporter, H as HangingProcessReporter, J as JUnitReporter, c as JsonReporter, M as MinimalReporter, R as ReportersMap, T as TapFlatReporter, d as TapReporter, V as VerboseBenchmarkReporter, e as VerboseReporter } from './chunks/index.UpGiHP7g.js';
import 'node:fs';
import './@vitest/runner/utils.js';
import './vendor/pathe.mjs';
import './vendor/tinyrainbow.mjs';
import './chunks/utils.BS4fH3nR.js';
import 'node:util';
import './@vitest/utils/helpers.js';
import 'node:fs/promises';
import 'node:perf_hooks';
import './@vitest/utils/source-map.js';
import './chunks/env.D4Lgay0q.js';
import './vendor/std-env.mjs';
import 'node:console';
import 'node:stream';
import './@vitest/utils/display.js';
import 'node:os';
import './vendor/tinyexec.mjs';
import './path.js';
import 'node:path';
import 'node:url';
import '@voidzero-dev/vite-plus-core';
import './@vitest/utils/offset.js';
import 'node:module';

console.warn("Importing from \"vitest/reporters\" is deprecated since Vitest 4.1. Please use \"vitest/node\" instead.");
